src/mlpack/methods/ann/vanilla_rbm_impl.hpp

```diff
@@ -0,0 +1,232 @@
+  SoftplusFunction::Fn(output, output);
+  return  -(arma::accu(output) + arma::dot(input, visible.Bias()));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Evaluate(const arma::mat& /* parameters*/, const size_t i)
+{
+  double freeEnergy, freeEnergyChain;
+  arma::mat negativeSample;
+  arma::vec currentInput = predictors.unsafe_col(i);
+  // Do not use unsafe col predictor as we change the input
+  freeEnergy = FreeEnergy(currentInput);
+  Gibbs(currentInput, std::move(negativeSample));
+  freeEnergyChain = FreeEnergy(negativeSample);
+
+  return arma::mean(freeEnergy) - arma::mean(freeEnergyChain);
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    MonitoringCost(const size_t i)
+{
+  arma::mat corruptInput, output;
+  corruptInput = predictors.col(i);
+  if(persistence)
+  {
+    arma::mat corruptFe, inputFe;
+    size_t idx = RandInt(0, predictors.n_rows);
+    corruptInput.row(idx) = 1 - corruptInput.row(idx);
+    // Use mean when showing the if more than one input
+    return std::log(LogisticFunction::Fn(FreeEnergy(corruptInput) - 
+        FreeEnergy(predictors.col(i)))) * predictors.n_rows;
+  }
+  else
+  {
+    visible.Forward(std::move(predictors.col(i)), std::move(output));
+    // Use mean when showing the if more than one input
+    return arma::mean(arma::accu( predictors.col(i) * arma::log(output).t() + 
+        (1 - predictors.col(i)) * arma::log(1 - output).t()));
+  }
+
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    SampleHidden(arma::mat&& input, arma::mat&& output)
+{
+  visible.Sample(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    SampleVisible(arma::mat&& input, arma::mat&& output)
+{
+  hidden.Sample(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Gibbs(arma::mat input, arma::mat&& output, size_t steps)
+{
+  steps = (steps == 1)?this-> numSteps: steps;
+  if (persistence && !state.is_empty())
+    input = state;
+
+  for (size_t j = 0; j < steps; j++)
+  {
+    // Use probabilties for updation till the last step(section 3 hinton)
+    ForwardVisible(std::move(input), std::move(output));
+    ForwardHidden(std::move(output), std::move(input));
+  }
+  // Set state to binarised output
+  SampleVisible(std::move(output), std::move(state));
```

Comment:
Are you getting better results with this strategy i am not getting better results with using probabilities?